Handle unknown project types

Adding a chat room of the "Talk" app to an existing project breaks the sync:
```
it.niedermann.nextcloud.deck.exceptions.TraceableException: Sorry, a wild error appeared!

⚠️ If you want to tell us about the following issue, please make sure to censor sensitive data beforehand! ⚠️

Failed to run traceable code with arguments:
{"ocs":{"meta":{"status":"ok","statuscode":200,"message":"OK"},"data":[{"id":3,"name":"Project 1","resources":[{"type":"deck-card","id":"605","name":"Artur is great!: llkkll","link":"/index.php/apps/deck/#/board/98/card/605","iconUrl":"/core/img/actions/toggle-pictures.svg"},{"type":"file","id":"46589","name":"2019_Glocken.mp4","path":"files/2019_Glocken.mp4","link":"https://example.com/index.php/f/46589","mimetype":"video/mp4","preview-available":false},{"type":"room","id":"so763ats","name":"sample","call-type":"group","iconUrl":"https://example.com/apps/spreed/img/app-dark.svg","link":"https://example.com/index.php/call/so763ats"}]},{"id":4,"name":"Sample Project","resources":[{"type":"deck","id":"135","name":"BoardWithoutAnything","link":"/index.php/apps/deck/#/board/135","iconUrl":"/apps/deck/img/deck-dark.svg"},{"type":"deck-card","id":"605","name":"Artur is great!: llkkll","link":"/index.php/apps/deck/#/board/98/card/605","iconUrl":"/core/img/actions/toggle-pictures.svg"}]},{"id":5,"name":"Another project","resources":[{"type":"deck-card","id":"558","name":"Deck Android: Test","link":"/index.php/apps/deck/#/board/99/card/558","iconUrl":"/core/img/actions/toggle-pictures.svg"},{"type":"deck-card","id":"605","name":"Artur is great!: llkkll","link":"/index.php/apps/deck/#/board/98/card/605","iconUrl":"/core/img/actions/toggle-pictures.svg"}]},{"id":6,"name":"llkkll","resources":[{"type":"deck","id":"135","name":"BoardWithoutAnything","link":"/index.php/apps/deck/#/board/135","iconUrl":"/apps/deck/img/deck-dark.svg"},{"type":"deck-card","id":"605","name":"Artur is great!: llkkll","link":"/index.php/apps/deck/#/board/98/card/605","iconUrl":"/core/img/actions/toggle-pictures.svg"}]},{"id":7,"name":"Project 2","resources":[{"type":"deck-card","id":"537","name":"Artur is great!: trester","link":"/index.php/apps/deck/#/board/98/card/537","iconUrl":"/core/img/actions/toggle-pictures.svg"},{"type":"deck-card","id":"605","name":"Artur is great!: llkkll","link":"/index.php/apps/deck/#/board/98/card/605","iconUrl":"/core/img/actions/toggle-pictures.svg"},{"type":"file","id":"6963","name":"Gnome Wallpaper.jpg","path":"files/Dokumente/Gnome Wallpaper.jpg","link":"https://example.com/index.php/f/6963","mimetype":"image/jpeg","preview-available":true}]}]}}
Cause: Sorry, a wild error appeared!

⚠️ If you want to tell us about the following issue, please make sure to censor sensitive data beforehand! ⚠️

Failed to run traceable code with arguments:
{"type":"room","id":"so763ats","name":"sample","call-type":"group","iconUrl":"https://example.com/apps/spreed/img/app-dark.svg","link":"https://example.com/index.php/call/so763ats"}
Cause: For input string: "so763ats"
	at it.niedermann.nextcloud.deck.api.JsonToEntityParser.parseOcsProjectList(JsonToEntityParser.java:97)
	at it.niedermann.nextcloud.deck.api.JsonToEntityParser.parseJsonObject(JsonToEntityParser.java:65)
	at it.niedermann.nextcloud.deck.api.NextcloudDeserializer.deserialize(NextcloudDeserializer.java:30)
	at com.google.gson.internal.bind.TreeTypeAdapter.read(TreeTypeAdapter.java:69)
	at com.google.gson.Gson.fromJson(Gson.java:932)
	at com.google.gson.Gson.fromJson(Gson.java:897)
	at com.nextcloud.android.sso.api.NextcloudAPI.performRequest(NextcloudAPI.java:101)
	at com.nextcloud.android.sso.api.NextcloudAPI.lambda$performRequestObservable$0$NextcloudAPI(NextcloudAPI.java:86)
	at com.nextcloud.android.sso.api.-$$Lambda$NextcloudAPI$yhhE-DMCe81fXF-bhjup5Tra-6o.subscribe(Unknown Source:6)
	at io.reactivex.internal.operators.observable.ObservableFromPublisher.subscribeActual(ObservableFromPublisher.java:31)
	at io.reactivex.Observable.subscribe(Observable.java:12284)
	at io.reactivex.internal.operators.observable.ObservableSubscribeOn$SubscribeTask.run(ObservableSubscribeOn.java:96)
	at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:578)
	at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
	at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:919)
Caused by: it.niedermann.nextcloud.deck.exceptions.TraceableException: Sorry, a wild error appeared!

⚠️ If you want to tell us about the following issue, please make sure to censor sensitive data beforehand! ⚠️

Failed to run traceable code with arguments:
{"type":"room","id":"so763ats","name":"sample","call-type":"group","iconUrl":"https://example.com/apps/spreed/img/app-dark.svg","link":"https://example.com/index.php/call/so763ats"}
Cause: For input string: "so763ats"
	at it.niedermann.nextcloud.deck.api.JsonToEntityParser.parseOcsProjectResource(JsonToEntityParser.java:131)
	at it.niedermann.nextcloud.deck.api.JsonToEntityParser.lambda$parseOcsProjectList$1(JsonToEntityParser.java:113)
	at it.niedermann.nextcloud.deck.api.-$$Lambda$JsonToEntityParser$QCh1Pf-C4m8nxLa8sWB94eGRxtA.run(Unknown Source:4)
	at it.niedermann.nextcloud.deck.exceptions.TraceableException.makeTraceableIfFails(TraceableException.java:13)
	... 20 more
Caused by: java.lang.NumberFormatException: For input string: "so763ats"
	at java.lang.Long.parseLong(Long.java:594)
	at java.lang.Long.parseLong(Long.java:636)
	at it.niedermann.nextcloud.deck.api.JsonToEntityParser.lambda$parseOcsProjectResource$2(JsonToEntityParser.java:132)
	at it.niedermann.nextcloud.deck.api.-$$Lambda$JsonToEntityParser$frO1oWvHO9S9GYM74W9rpTjE4bU.run(Unknown Source:4)
	at it.niedermann.nextcloud.deck.exceptions.TraceableException.makeTraceableIfFails(TraceableException.java:13)
	... 24 more

```

--- a/app/src/main/java/it/niedermann/nextcloud/deck/api/JsonToEntityParser.java
+++ b/app/src/main/java/it/niedermann/nextcloud/deck/api/JsonToEntityParser.java
@@ -129,7 +129,16 @@ public class JsonToEntityParser {
         DeckLog.verbose(obj.toString());
         OcsProjectResource resource = new OcsProjectResource();
         makeTraceableIfFails(() -> {
-            resource.setId(Long.parseLong(getNullAsEmptyString(obj.get("id"))));
+            if (obj.has("id")) {
+                String idString = obj.get("id").getAsString();
+                if (idString != null && idString.trim().length() > 0) {
+                    if (idString.matches("[0-9]+")){
+                        resource.setId(Long.parseLong(idString.trim()));
+                    } else {
+                        resource.setIdString(idString);
+                    }
+                }
+            }
             if (obj.has("type")) {
                 resource.setType(getNullAsEmptyString(obj.get("type")));
             }

--- a/app/src/main/java/it/niedermann/nextcloud/deck/model/ocs/projects/OcsProjectResource.java
+++ b/app/src/main/java/it/niedermann/nextcloud/deck/model/ocs/projects/OcsProjectResource.java
@@ -16,7 +16,7 @@ import it.niedermann.nextcloud.deck.model.interfaces.AbstractRemoteEntity;
         indices = {
                 @Index(value = "id", name = "index_OcsProjectResource_id"),
                 @Index(value = "lastModifiedLocal", name = "index_OcsProjectResource_lastModifiedLocal"),
-                @Index(value = {"accountId", "id", "projectId"}, name = "index_OcsProjectResource_accountId_id", unique = true),
+                @Index(value = {"accountId", "id", "idString", "projectId"}, name = "index_OcsProjectResource_accountId_id", unique = true),
                 @Index(value = "accountId", name = "index_projectResource_accID"),
                 @Index(value = "projectId", name = "index_projectResource_projectId"),
         },
@@ -43,6 +43,9 @@ public class OcsProjectResource extends AbstractRemoteEntity implements Serializ
     private String mimetype;
     @Nullable
     private Boolean previewAvailable;
+    @Nullable
+    private String idString;
+
 
     @NonNull
     private Long projectId;
@@ -119,5 +122,14 @@ public class OcsProjectResource extends AbstractRemoteEntity implements Serializ
 
     public void setPreviewAvailable(@Nullable Boolean previewAvailable) {
         this.previewAvailable = previewAvailable;
+    }
+
+    @Nullable
+    public String getIdString() {
+        return idString;
+    }
+
+    public void setIdString(@Nullable String idString) {
+        this.idString = idString;
     }
 }

--- a/app/src/main/java/it/niedermann/nextcloud/deck/persistence/sync/adapters/db/DeckDatabase.java
+++ b/app/src/main/java/it/niedermann/nextcloud/deck/persistence/sync/adapters/db/DeckDatabase.java
@@ -87,7 +87,7 @@ import it.niedermann.nextcloud.deck.persistence.sync.adapters.db.dao.widgets.Sta
                 JoinCardWithProject.class,
         },
         exportSchema = false,
-        version = 18
+        version = 19
 )
 @TypeConverters({DateTypeConverter.class})
 public abstract class DeckDatabase extends RoomDatabase {
@@ -227,6 +227,15 @@ public abstract class DeckDatabase extends RoomDatabase {
             database.execSQL("ALTER TABLE `Account` ADD `etag` TEXT");
         }
     };
+    private static final Migration MIGRATION_18_19 = new Migration(18, 19) {
+        @Override
+        public void migrate(SupportSQLiteDatabase database) {
+            // https://github.com/stefan-niedermann/nextcloud-deck/issues/619
+            database.execSQL("DROP INDEX `index_OcsProjectResource_accountId_id`");
+            database.execSQL("ALTER TABLE `OcsProjectResource` ADD `idString` TEXT");
+            database.execSQL("CREATE UNIQUE INDEX `index_OcsProjectResource_accountId_id` ON `OcsProjectResource` (`accountId`, `id`, `idString`, `projectId`)");
+        }
+    };
 
     public static final RoomDatabase.Callback ON_CREATE_CALLBACK = new RoomDatabase.Callback() {
         @Override
@@ -271,6 +280,7 @@ public abstract class DeckDatabase extends RoomDatabase {
                 .addMigrations(MIGRATION_15_16)
                 .addMigrations(MIGRATION_16_17)
                 .addMigrations(MIGRATION_17_18)
+                .addMigrations(MIGRATION_18_19)
                 .fallbackToDestructiveMigration()
                 .addCallback(ON_CREATE_CALLBACK)
                 .build();

--- a/fastlane/metadata/android/en-US/changelogs/1009003.txt
+++ b/fastlane/metadata/android/en-US/changelogs/1009003.txt
@@ -1,0 +1,1 @@
+- 🐞 Handle project resources with a string id properly

